Update copyright headers

ptagl opened the pull request:
This pull request introduces the following changes:
- a new Python script `update-copyright-headers.py` to update copyright headers in the source files of the repository
- updated headers for "Zen Blockchain Foundation" for all the files modified between `2018-01-18` and `2023-09-27`
- updated headers for "The Horizen Foundation" for all the files modified between `2023-09-28` and `2024-05-19`

The previous script `fix-copyright-headers.py` has been removed and the `devtools` readme has been updated accordingly.

Commit:
Added a new Python script for copyright headers update

The new script "update-copyright-headers.py" replaces the previous "fix-copyright-headers.py".

The script can be run to update an existing header and also add new ones.

It contains an help message explaining all the input parameters that can be used to tune the output.

Note that the script checks for required updates by inspecting the Git history and check if any file was modified outside the copyright range specified in the header (e.g. the copyright states 2018-2020 but it finds a change in Git history on 2021, in which case the header is updated).

ptagl commented on contrib/devtools/README.md:
```diff
@@ -41,15 +41,19 @@ Configuring the github-merge tool for the bitcoin repository is done in the foll
-Every year newly updated files need to have its copyright headers updated to reflect the current year.
-If you run this script from src/ it will automatically update the year on the copyright header for all
-.cpp and .h files if these have a git commit from the current year.
+This scripts takes care of updating the copyright headers in the source files of the repository.
```
```suggestion
This script takes care of updating the copyright headers in the source files of the repository.
```

ptagl commented on contrib/devtools/README.md:
```diff
@@ -41,15 +41,19 @@ Configuring the github-merge tool for the bitcoin repository is done in the foll
 fix-copyright-headers.py
```
```suggestion
update-copyright-headers.py
```

Commit:
Updated the "Zen Blockchain Foundation" header

The changes have been automatically produced by running:

`python contrib/devtools/update-copyright-headers.py --since 2018-
01-18 --until 2023-09-27 --copyright-holder "Zen Blockchain Foundation"`

Commit:
Added "The Horizen Foundation" copyright header to source files

Only files modified since 2023-09-28 are included.

The changes have been automatically generated with:

`python contrib/devtools/update-copyright-headers.py --since 2023-
09-28 --until 2024-05-19 --copyright-holder "The Horizen Foundation"`

Note that the "--until" parameter was important to avoid including files updated with the previous commit!

drgora reviewed (approved):
Looks good to me, thanks!

JackPiri commented on contrib/devtools/update-copyright-headers.py:
```diff
@@ -0,0 +1,188 @@
+    """
+    file_list = []
+    for root, dirs, files in os.walk(path):
+        # Exclude specified subfolders
```
From this code I understand the `args.exclude_dir` set is targeting only the first level of `args.directory` (not deeper subdirectories):

```
    # Check if the specified excluded subfolders exist
    for folder in args.exclude_dir:
        path = os.path.join(args.directory, folder)
        if not os.path.exists(path):
            print(f"Error: Folder '{path}' does not exist in the specified directory.")
            exit(1)
```

if my understanding is correct, probably we want to modify a little bit `list_files` function:

```
    file_list = []
    first_level = True
    for root, dirs, files in os.walk(path):
        if first_level:
            # Exclude specified subfolders
            dirs[:] = [d for d in dirs if d not in exclude_folders]
            first_level = False
        for file in files:
```

in the situation we have deeper subdirectories named the same way (not the case as of now).

ptagl replied:
Thanks for highlighting this, I improved the script by making the handling of paths more robust, combining relative and absolute paths.

This should now guarantee even more flexibility without the risk of multiple matches for the same folder.
In particular, supposing to have a folder tree like:
```
src/
├─ A/
├─ B/
├─ C/
│  ├─ A/
```
excluding the folder `A` wouldn't also cause `C/A` to be excluded, as it should be provided as a relative path to `src`.

JackPiri replied:
It's fine to me now, thanks.

JackPiri reviewed (approved):
The PR is fine to me, I've left just a minor comment for making the script a bit more coherent.
Thanks

Commit:
Restored f-string for pattern matching

F-strings and r-strings can be mixed, but characters like "{" and "}" need to be escaped properly.

Commit:
Improved management of excluded folders

The previous version could exclude subfolders in case they had the same name of any folder at the top level.
The new version uses absolute paths to be more specific.

Commit:
Added the possibility to exclude some commits

The script now includes the "--exclude-commit" parameter that can be used to specify commits that we don't want to track for file changes.

This is extremely useful when there are changes that shouldn't affect the copyright header, for instance the update of the copyright header itself!

By default, the argument is an empty list, in which case the script tries to guess the commits to be excluded by looking for the presence of the string "update-copyright-headers.py" in the message of the commit or "copyright" in the summary.

If users explicitly sets the argument, no automatic detection is performed.

Note that in case of automatic detection, the script asks user for confirmation before proceeding.

Commit:
Fixed typo in the help message of update-copyright-header.py

Commit:
Improved detection of excluded commits for update-copyright-headers.py

The check over the commit message is not case-sensitive anymore.
